### PR TITLE
[expo-cli] Add a eas:build:init command

### PR DESCRIPTION
**Pull request by satya164**
## Why

We want to add a separate command to configure the project for eas builds without having to actually start a build.

## Test plan

Run `expo eas:build:init` in a project. I had some issues with generating credentials on iOS but it's unrelated to this PR as I haven't changed any logic for credentials.

Kapture 2020-08-25 at 20 13 45

### CODE REVIEW

File: packages/expo-cli/src/commands/eas-build/init/action.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+  const { platform = BuildCommandPlatform.ALL, profile } = options;
+
+  if (!platforms.includes(platform)) {
+    throw new Error(
+      `-p/--platform needs a valid platform: ${platforms.map(p => log.chalk.bold(p)).join(', ')}`
+    );
+  }
+
+  const spinner = ora('Creating minimal eas.json file');
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/eas-build/init/action.ts:39]: make sure you always succeed or fail on that spinner

File: packages/expo-cli/src/commands/eas-build/init/action.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+interface BuildOptions {
+  platform: BuildCommandPlatform;
+  skipCredentialsCheck?: boolean; // TODO: noop for now
+  profile: string;
+  parent?: {
+    nonInteractive: boolean;
+  };
+}
+
+async function initAction(projectDir: string, options: BuildOptions): Promise<void> {
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/eas-build/init/action.ts:28]: I'm not 100% sure what is commanderjs behavior where, but from what I remember if you run
`expo eas:init --profile` then this action would be triggered with options set to  `{ profile: true }`

Make sure if I'm correct and handle that case if necessary

File: packages/expo-cli/src/commands/eas-build/init/action.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  if (!platforms.includes(platform)) {
+    throw new Error(
+      `-p/--platform needs a valid platform: ${platforms.map(p => log.chalk.bold(p)).join(', ')}`
+    );
+  }
+
+  const spinner = ora('Creating minimal eas.json file');
+
+  ensureGitStatusIsCleanAsync();
```
Comment (dsokal) [inline packages/expo-cli/src/commands/eas-build/init/action.ts:41]:
```suggestion
  await ensureGitStatusIsCleanAsync();
```

File: packages/expo-cli/src/commands/eas-build/build/action.ts [hunk 4]
```diff
@@ -145,7 +111,13 @@ async function startBuildAsync(
...
 
     if (!builder.ctx.skipProjectConfiguration) {
-      await builder.configureProjectAsync();
+      if (builder instanceof iOSBuilder) {
+        await builder.configureProjectAsync();
+      } else if (!(await builder.isProjectConfiguredAsync())) {
+        throw new CommandError(
+          'Project is not configured. Please run "expo eas:build:init" first to configure the project'
+        );
+      }
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/eas-build/build/action.ts:120]:
```suggestion
      await builder.enureProjectConfiguredAsync();
```
and `enureProjectConfiguredAsync` on android would  contain only
```
     if (!(await this.isProjectConfiguredAsync())) {
        throw new CommandError(
          'Project is not configured. Please run "expo eas:build:init" first to configure the project'
        );
      }
```

I would suggest replacing `configureProjectAsync` -> `enureProjectConfiguredAsync` in the builder interface

configureProjectAsync would be a public method on builders but not the part of the interface

when building you would call `enureProjectConfiguredAsync` methods
when running init you can call directly `configureProjectAsync`

File: packages/expo-cli/src/commands/eas-build/init/action.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+  const ctx = await createBuilderContextAsync(projectDir, easConfig, {
+    platform: BuildCommandPlatform.ALL,
+    nonInteractive,
+    skipCredentialsCheck: options?.skipCredentialsCheck,
+    skipProjectConfiguration: false,
+  });
+
+  const androidBuilder = new AndroidBuilder(ctx);
+
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/eas-build/init/action.ts:94]: it's not necessary on android, but if we are forcing the user to setup ios credentials we should also setup android part

**Review by wkozyra95 — CHANGES_REQUESTED** [review]

File: packages/expo-cli/src/commands/eas-build/build/action.ts (hunk 4 above)
Comment (wkozyra95) [reply packages/expo-cli/src/commands/eas-build/build/action.ts:120]: as a side note
we will probably want to remove project configuration and credentials setup from the builder in the future, this interface should be kept as small as posible

**Review by wkozyra95 — APPROVED** [review]

**Review by dsokal — APPROVED** [review]

**Comment by dsokal** [comment]
Looks good, please update the changelog! :)

**Comment by wkozyra95** [comment]
Also, remember to update documentation here https://github.com/expo/expo/tree/master/docs/pages/build

**Comment by dsokal** [comment]
> Also, remember to update documentation here https://github.com/expo/expo/tree/master/docs/pages/build

Yeah, actually please open a PR to docs before merging this. 
https://github.com/expo/expo-cli/pull/2442 - please update the docs for this feature too

**Comment by wkozyra95** [comment]
> > Also, remember to update documentation here https://github.com/expo/expo/tree/master/docs/pages/build
> 
> Yeah, actually please open a PR to docs before merging this.
> #2442 - please update the docs for this feature too

We talked offline with dominik and we think that it's better to merge it earlier, just make sure that docs are ready before expo-cli is published.

**Pull request merged**